src/modules/landing_target_estimator/LandingTargetEstimator.cpp

```diff
@@ -31,279 +31,307 @@
 
-	} else {
-		// update
-		const float measurement_uncertainty = _params.meas_unc * _dist_z * _dist_z;
-		bool update_x = _kalman_filter_x.update(_target_position_report.rel_pos_x, measurement_uncertainty);
-		bool update_y = _kalman_filter_y.update(_target_position_report.rel_pos_y, measurement_uncertainty);
-
-		if (!update_x || !update_y) {
-			if (!_faulty) {
-				_faulty = true;
-				PX4_WARN("Landing target measurement rejected:%s%s", update_x ? "" : " x", update_y ? "" : " y");
+			if (!vehicle_local_position.v_xy_valid || !vehicle_local_position.dist_bottom_valid
+			    || !_vehicle_attitude_sub.update(&vehicle_attitude)
+			    || !PX4_ISFINITE(irlock_report.pos_y) || !PX4_ISFINITE(irlock_report.pos_x)) {
+				// don't have the data needed for an update
+				return;
 			}
 
-		} else {
-			_faulty = false;
-		}
+			matrix::Vector3f sensor_ray; // ray pointing towards target in body frame
+			sensor_ray(0) = irlock_report.pos_x * _param_ltest_scale_x.get(); // forward
+			sensor_ray(1) = irlock_report.pos_y * _param_ltest_scale_y.get(); // right
+			sensor_ray(2) = 1.f;
 
-		if (!_faulty) {
-			// only publish if both measurements were good
+			// Orientation of the sensor relative to body frame
+			const matrix::Dcmf S_att = get_rot_matrix((Rotation)_param_ltest_sens_rot.get());
+			// rotate unit ray according to sensor orientation
+			sensor_ray = S_att * sensor_ray;
+
+			// rotate the unit ray into the navigation frame
+			const matrix::Quatf q_att(vehicle_attitude.q);
+			const matrix::Dcmf R_att(q_att); // Orientation of the body frame
+			sensor_ray = R_att * sensor_ray;
 
-			_target_pose.timestamp = _target_position_report.timestamp;
+			if (fabsf(sensor_ray(2)) < 1e-6f) {
+				// z component of measurement unsafe, don't use this measurement
+				return;
+			}
 
-			float x, xvel, y, yvel, covx, covx_v, covy, covy_v;
-			_kalman_filter_x.getState(x, xvel);
-			_kalman_filter_x.getCovariance(covx, covx_v);
+			matrix::Vector3f rel_pos{};
 
-			_kalman_filter_y.getState(y, yvel);
-			_kalman_filter_y.getCovariance(covy, covy_v);
+			const float dist_z = vehicle_local_position.dist_bottom - _param_ltest_sens_pos_z.get();
 
-			_target_pose.is_static = (_params.mode == TargetMode::Stationary);
+			// scale the ray s.t. the z component has length of _uncertainty_scale
+			//  - adjust relative position according to sensor offset
+			rel_pos(0) = (sensor_ray(0) / sensor_ray(2) * dist_z) + _param_ltest_sens_pos_x.get();
+			rel_pos(1) = (sensor_ray(1) / sensor_ray(2) * dist_z) + _param_ltest_sens_pos_x.get();
+			rel_pos(2) = dist_z;
 
-			_target_pose.rel_pos_valid = true;
-			_target_pose.rel_vel_valid = true;
-			_target_pose.x_rel = x;
-			_target_pose.y_rel = y;
-			_target_pose.z_rel = _target_position_report.rel_pos_z ;
-			_target_pose.vx_rel = xvel;
-			_target_pose.vy_rel = yvel;
+			if (!_estimator_initialized) {
+				float vx_init = -vehicle_local_position.vx;
+				float vy_init = -vehicle_local_position.vy;
```

Comment:
```suggestion
				const float vx_init = -vehicle_local_position.vx;
				const float vy_init = -vehicle_local_position.vy;
```

This might require adding `const` to the KalmanFilter::init() function as well